proc/redis/filter_compress.go

```diff
@@ -0,0 +1,323 @@
+)
+
+/*
+Transparent compression is to reduce the use of memory and
+network bandwidth by big keys and improve request performance.
+Mark the data to be compressed by adding a header before the
+compressed data, the following diagram depicts the protocol:
+
+                                                +
+                                                |
+                     header                           compressed data
+                                                |
+ 1byte                                          |
++------+-------+-------+-------+--------+----------------------------------+
+|      |       |       |       |        |       |         ...
++--+---+-------+--+----+---+---+----+---+--+-------------------------------+
+   |              |        |        |      |    |
+   |              |    algorithm    +--+---+    |
+   +-------+------+                    +        |
+           +                         \r\n       |
+      magic number                              |
+                                                |
+                                                +
+*/
+
+// MagicNumber & Separator is set for the Header
+const (
+	MagicNumber = "(P$"
+	Separator   = "\r\n"
+)
+
+var (
+	cpsHdrLen = len(MagicNumber) + 1 + len(Separator)
+
+	errMissingCpsHdr          = errors.New("missing cps header")
+	errCpsMagicNumberNotFound = errors.New("cps magic number not found")
+	errInvalidCpsType         = errors.New("invalid cps type")
+	errUnsupportedCpsType     = errors.New("unsupported cps type")
+
+	// bannedCmdInCps is a set of commands that are disabled in compress mode.
+	bannedCmdInCps = map[string]struct{}{
+		"append":   {},
+		"eval":     {},
+		"setbit":   {},
+		"getbit":   {},
+		"setrange": {},
+		"getrange": {},
+	}
+
+	// well known commands which could skip decompress check
+	wkSkipCheckCmdsInDecps = map[string]struct{}{
+		// string commands
+		"incr":   {},
+		"decr":   {},
+		"incrby": {},
+		"decrby": {},
+
+		// Hash commands
+		"hkeys":        {},
+		"hincrby":      {},
+		"hincrbyfloat": {},
+
+		// List commands
+		"lpop":   {},
+		"lpush":  {},
+		"rpop":   {},
+		"rpush":  {},
+		"lrange": {},
+		"lindex": {},
+
+		// Set commands
+		"spop":     {},
+		"sunion":   {},
+		"smembers": {},
+
+		// Zset commands
+		"zadd":          {},
+		"zrange":        {},
+		"zcard":         {},
+		"zrangebyscore": {},
+
+		"cluster": {},
+		"ping":    {},
+	}
+
+	cpsHdrs = map[redis.Compression_Method][]byte{}
+)
+
+func init() {
+	methods := []redis.Compression_Method{snappy.Name}
+	for _, method := range methods {
+		hdr := make([]byte, cpsHdrLen)
+		copy(hdr, MagicNumber)
+		hdr[len(MagicNumber)+1] = byte(method)
+		copy(hdr[len(MagicNumber)+2:], Separator)
+		cpsHdrs[method] = hdr
+	}
+}
+
+// writeCpsHeader generates compress header from specific compressType
+func writeCpsHeader(typ redis.Compression_Method, buf buffer) {
+	if cpsHdr, ok := cpsHdrs[typ]; ok {
+		buf.Write(cpsHdr)
+	} else {
+		// for test
```

Comment:
It should return an error, maybe the configured compress method is wrong or unsupported